scertecd/scertecd.go

```diff
@@ -660,12 +672,137 @@ func (cu *certUpdateCheck) updateIfNeeded(ctx context.Context, prev *certUpdateR
+		return fmt.Errorf("failed to PEM encode cert signed by root %q: %w", cu.s.ActiveRoot, err)
+	}
+
+	// ensure that we can parse the resulting key/cert bundle.
+	m, err := cu.s.parseCertMeta(allPEM.Bytes(), true)
+	if err != nil {
+		cu.s.addError("private-renewal-parse-new-cert")
+		return fmt.Errorf("failed to parse newly fetched cert: %w", err)
+	}
+
+	cu.res.SecretVersion, err = cu.s.putAndActivateSecret(ctx, cu.Logf, secName, allPEM.Bytes())
+	if err != nil {
+		return err
+	}
+	cu.res.Updated = true
+	cu.res.CertMeta = m
+	return nil
+}
+
+// signWithRoot signs newCert with the root CA identified by rootPrivKeySecName,
+// using newKey as the private key for newCert. The returned byte slice contains
+// the DER-encoded signed certificate.
+func (cu *certUpdateCheck) signWithRoot(ctx context.Context, rootPrivKeySecName string, newCert *x509.Certificate, newKey crypto.Signer) ([]byte, error) {
+	// retrieve root private key; PEM decode & parse it
+	caPrivKeyName := path.Join(cu.s.Prefix, "roots", "private-key")
+	caPrivKeySecret, err := cu.s.getSecret(ctx, caPrivKeyName)
+	if err != nil {
+		return nil, fmt.Errorf("getSecret %q: %w", caPrivKeyName, err)
+	}
+	keyPB, rest := pem.Decode(caPrivKeySecret.Value)
+	if keyPB == nil {
+		return nil, fmt.Errorf("pem.Decode %q: no PEM data found", caPrivKeyName)
+	}
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("pem.Decode: %q: unexpected trailing data after key PEM block", caPrivKeyName)
+	}
+	caPrivKey, err := x509.ParsePKCS8PrivateKey(keyPB.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse private key %q: %v", caPrivKeyName, err)
+	}
+
+	// retrieve root certificate; PEM decode & parse it
+	caCertName := path.Join(cu.s.Prefix, "roots", rootPrivKeySecName, "certificate-latest")
+	caCertSecret, err := cu.s.getSecret(ctx, caCertName)
+	if err != nil {
+		return nil, fmt.Errorf("getSecret %q: %w", caCertName, err)
+	}
+	certPB, _ := pem.Decode(caCertSecret.Value)
+	if certPB == nil {
+		return nil, fmt.Errorf("pem.Decode %q: no PEM data found", caCertName)
+	}
+	caCert, err := x509.ParseCertificate(certPB.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("parsePEMCertificates %q: %w", caCertName, err)
+	}
+
+	// sign the new certificate with the root private key
+	return x509.CreateCertificate(rand.Reader, newCert, caCert, newKey.Public(), caPrivKey)
+}
+
+func (cu *certUpdateCheck) newCert() (*x509.Certificate, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate serial number: %w", err)
+	}
+
+	return &x509.Certificate{
+		SerialNumber: n,
+		DNSNames:     []string{cu.dt.domain},
+		Subject: pkix.Name{
+			Organization: []string{"Tailscale, Inc."},
+			CommonName:   cu.dt.domain,
+		},
+		NotBefore:   cu.s.Now(),
+		NotAfter:    cu.s.Now().Add(30 * 24 * time.Hour), // valid for 30 days
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
```

Comment:
TODO to stop setting ExtKeyUsageClientAuth for public ACME certs? need to plumb that in?